README.md
# DriftBurstHypothesis

An R-package for the calculation of the Drift Burst Hypothesis test-statistic from the working paper Christensen, Oomen and Reno (2018) <DOI:10.2139/ssrn.2842535>.

The t-statistic at period n is calculated as follows:

![equation](https://latex.codecogs.com/png.latex?%5Cbar%7BT%7D%5En%20%3D%20%5Csqrt%7B%5Cfrac%7Bh_%7Bn%7D%7D%7BK_%7B2%7D%7D%7D%5Cfrac%7B%5Chat%7B%5Cbar%7B%5Cmu%7D%7D_%7Bt%7D%5E%7Bn%7D%7D%7B%5Csqrt%7B%5Chat%7B%5Cbar%7B%5Csigma%7D%7D_%7Bt%7D%5E%7Bn%7D%7D%7D), 

where the local mean estimator is:

![equation](https://latex.codecogs.com/png.latex?%5Chat%7B%5Cbar%7B%5Cmu%7D%7D_%7Bt%7D%5E%7Bn%7D%3D%5Cfrac%7B1%7D%7Bh_%7Bn%7D%7D%5Csum_%7Bi%3D1%7D%5E%7Bn-k_%7Bn%7D&plus;2%7DK%5Cleft%28%5Cfrac%7Bt_%7Bi-1%7D-t%7D%7Bh_%7Bn%7D%7D%5Cright%29%5CDelta_%7Bi-1%7D%5E%7Bn%7D%5Coverline%7BY%7D),

and the local variance estimator is:

![equation](https://latex.codecogs.com/png.latex?%5Chat%7B%5Cbar%7B%5Csigma%7D%7D_%7Bt%7D%5E%7Bn%7D%20%3D%20%5Cfrac%7B1%7D%7Bh_%7Bn%7D%27%7D%5Cleft%5B%5Csum_%7Bi%3D1%7D%5E%7Bn-k_%7Bn%7D&plus;2%7D%5Cleft%28K%5Cleft%28%5Cfrac%7Bt_%7Bi-1%7D-t%7D%7Bh%27_%7Bn%7D%7D%5Cright%29%5CDelta_%7Bi-1%7D%5E%7Bn%7D%5Coverline%7BY%7D%5Cright%29%5E%7B2%7D&plus;2%5Csum_%7BL%3D1%7D%5E%7BL_%7Bn%7D%7D%5Comega%5Cleft%28%5Cfrac%7BL%7D%7BL_%7Bn%7D%7D%5Cright%29%5Csum_%7Bi%3D1%7D%5E%7Bn-k_%7Bn%7D-L&plus;2%7DK%5Cleft%28%5Cfrac%7Bt_%7Bi-1%7D-t%7D%7Bh_%7Bn%7D%27%7D%5Cright%29K%5Cleft%28%5Cfrac%7Bt_%7Bi&plus;L-1%7D-t%7D%7Bh_%7Bn%7D%27%7D%5Cright%29%5CDelta_%7Bi-1%7D%5E%7Bn%7D%5Coverline%7BY%7D%5CDelta_%7Bi-1&plus;L%7D%5E%7Bn%7D%5Coverline%7BY%7D%5Cright%5D)


with:

![equation](https://latex.codecogs.com/png.latex?%5CDelta_%7Bi%7D%5E%7Bn%7D%5Coverline%7BY%7D%20%3D%20%5Csum_%7Bj%3D1%7D%5E%7Bk_%7Bn%7D-1%7Dg_%7Bj%7D%5E%7Bn%7D%5CDelta_%7Bi&plus;j%7D%5E%7Bn%7DY)

denoting the overlapping pre-averaged returns with the weighting function:


![equation](https://latex.codecogs.com/png.latex?g%5Cleft%28x%5Cright%29%3D%5Ctext%7Bmin%7D%5Cleft%28x%2C1-x%5Cright%29),

and

![equation](https://latex.codecogs.com/png.latex?%5Comega%5Cleft%28%5Ccdot%5Cright%29)

is a smooth kernel defined on the positive real numbers, ![equation](https://latex.codecogs.com/png.latex?L_%7Bn%7D) is the lag length over which the estimator is applied. By default, the lag-length will be determined by way of the Newey-West algorithm.





## Examples using simulated high frequency data:
```
library(highfrequency)
library(xts)
library(DriftBurstHypothesis)
data("sample_tdata")
price = xts(as.numeric(sample_tdata$PRICE), index(sample_tdata))
plot(price)
testtimes = seq(34260, 57600, 60)



DBHxts = drift_bursts(timestamps = NULL,  logpricexts,
                      testTimes, preAverage = 5, ACLag = -1L,
                      meanBandwidth = 300L, varianceBandwidth = 900L,
                      bParallelize = TRUE, iCores = 8)

plot(DBHxts, price = price)
```

![Example plot](https://i.imgur.com/LTWi7uM.png)


```
library(DriftBurstHypothesis)
set.seed(1234)
returns = rnorm(23399, sd = 1)/sqrt(23400)
price = c(0,cumsum(returns))
timestamps = seq(34200, 57600, length.out = 23400)
testTimes = seq(34200 + 5*300, 57600, 60)
DBH = driftBursts(timestamps, price, testTimes, preAverage = 5, meanBandwidth = 300, varianceBandwidth = 5*300)
plot(DBH, price = price, timestamps = timestamps)
```

![Example plot](https://i.imgur.com/m1v3ACN.png)


ChangeLog
2019-24-01 Emil Sjoerup <emilsjoerup@live.dk>
  *DESCRIPTION (Version): New version is 0.1.1
  *Fix ambiguity in a call to sqrt() for int which caused installation failure on Solaris.
  *More robust code, passing vectors of differing length for logprices and time are now stopped.
  *The Mu and Sigma vectors returned now also start with a 0 instead of uninitiated memory.
  *Documentation changed to make it more clear that the left-sided exponential kernel is used (no foresight).
2019-02-04 Emil Sjoerup <emilsjoerup@live.dk>
  *DESCRIPTION (Version): New version is 0.1.2
  *Fix a bug where the iMaxLag was being scaled higher if the lag selection algorithm suggested a lag higher than 20. (Was a bug now fixed)
  *Add xts funcitonality, now an xts object will no longer break the code and the timestamps can now be extracted from and xts.
2019-08-04 Emil Sjoerup <emilsjoerup@live.dk>
  *Mitigate bug that caused crashes if the first testing time was before actual observations.
2019-13-04 Emil Sjoerup <emilsjoerup@live.dk>
  *Update documentation to get ready for CRAN update.
2019-25-04 Emil Sjoerup <emilsjoerup@live.dk>
  *DESCRIPTION (Version): New version is 0.1.3
  *Implement tests for function that is prone to cause failures (Automatic Lag Selection Algorithm)
  *Put in safeguard that should prevent crashes. Now, it is not allowed to have more than 15 minutes between trades.
  *Fix bug caused by a misunderstanding when re-reading the paper. The bug was introduced in version 0.1.2 caused too low maximum lags
2019-14-05 Emil Sjoerup <emilsjoerup@live.dk>
  *New plot method for the "DBH" S3 class also introduced in this version
  *New methods getMu and getSigma to get the mu and sigma estimates
2019-06-06 Emil Sjoerup <emilsjoerup@live.dk>
  *New method getDB to get the Drift Bursts test statistic added.
2019-06-14 Emil Sjoerup <emilsjoerup@live.dk>
  *add timezones to the code, output should no longer have timezones different from the input
2019-06-24 Emil Sjoerup <emilsjoerup@live.dk>
  *Fix a mistake that lead to foresight in the lag selection algorithm
  *More helpful messages
2019-06-25 Emil Sjoerup <emilsjoerup@live.dk>
  *Allow for different time-zones in plotting. The standard is that a trading day starts at 9:30
2019-06-30 Emil Sjoerup <emilsjoerup@live.dk>
  *Use warnings instead of cat() to display warnings
  *Start loop at first test time instead of second in case users don't use same testtimes as I do.
  *getMu() and getSigma() now returns corrected values (i.e. mu * Mean_bandwidth, and sigma * 2 * iT)
2019-07-06 Emil Sjoerup <emilsjoerup@live.dk>
  *Bump version number to 0.1.3.1
  *Don't start loop at first test time, instead make it explicit in the documentation that the first test is not calculated and should be used to designate the start of the trading day.
  *Add tests for the examples given.
2019-08-22 Emil Sjoerup <emilsjoerup@live.dk>
  *Change to camelCase
  *getMu and getSigma now corrects for nObs
2019-09-01 Emil Sjoerup <emilsjoerup@live.dk>
  *DESCRIPTION (Version): New version is 0.2.0
  *Add realTime argument to driftBursts function
2019-10-02 Emil Sjoerup <emilsjoerup@live.dk>
  *No longer calling -abs(vX) in DBH.cpp as it was always a negative vector and was leftover from before I optimized indexing in that function. -Minor performance gain.
  *add print.DBH method
2019-10-06 Emil Sjoerup <emilsjoerup@live.dk>
  *DESCRIPTION (Version): New version is 0.2.0.1
  *Add .0 to 2.0 in floor(iLag/2) to floor(iLag/2.0) in order to avoid integer arguments in floor() which causes installation errors on Solaris
2019-10-27 Emil Sjoerup <emilsjoerup@live.dk>
  *Remove surperfluous code form the cfilter function - extra unused arguments were included, which have been removed.
  *Add padding vector to the output of driftBursts function, which is used for the new mean and variance methods of the new classes for the 3 estimates
  *Add methods for calculating mean and variance of the output of driftBursts
  *Re-organize the documentation to have a file for the methods of the DBH class.
2019-10-29 Emil Sjoerup <emilsjoerup@live.dk>
  *DESCRIPTION (Version): New version is 0.3.0 
  *In driftBursts, the testTimes argument should no longer have an entry denoting the start of the trading day, as now testing is done on all entries.
2019-11-23 Emil Sjoerup <emilsjoerup@live.dk>
  *Port over the routine for testing whether a flash crash is present from Matlab. 
2019-11-28 Emil Sjoerup <emilsjoerup@live.dk>
  *remove realTime feature, which was a bug mistaken as a feature :)


tests/testthat.R
library(testthat)
library(DriftBurstHypothesis)

test_check("DriftBurstHypothesis")


tests/testthat/test_DBH.R
library(testthat)
library(DriftBurstHypothesis)

context("DBH C++ error tests")
test_that("DBH C++ error tests",{
  #these functions has broken before, hopefully they won't break again.
  expect_equal(DriftBurstHypothesis:::AsymptoticVarianceC(c(1:3), 3), NaN) 
  expect_equal(DriftBurstHypothesis:::AsymptoticVarianceC(c(1:3), 4), NaN)
  expect_equal(DriftBurstHypothesis:::AutomaticLagSelectionC(1:10, 30) , 7)
  }
)

context("DBH test")
test_that("DBH sim test", {
  set.seed(123)
  iT = 23400; meanBandwidth = 300L
  timestamps = seq(0, 23400, length.out = iT+1)
  testtimes  = seq(60, 23400, 60L)
  
  r = rnorm(iT, mean = 0.02, sd = 1)/sqrt(iT)
  p = c(0,cumsum(r))
  
  
  DBH = driftBursts(timestamps, p, testtimes, preAverage = 1, ACLag = -1, 
                     meanBandwidth = meanBandwidth, varianceBandwidth = 5*meanBandwidth, parallelize = FALSE)
  #The new mu conversion is 1/2*OldMu
  expect_equal(mean(getMu(DBH)),0.01253808)
  expect_equal(mean(getSigma(DBH)), 0.9748568)
})


context("Examples check")
test_that("DBH Examples check",{
  #Both a flash crash and flash rally are coded into the function.
  flashCrashSim = function(iT, dSigma, dPhi, dMu){
    vSim = numeric(iT)
    vEps = rnorm(iT , sd =dSigma)
    vEpsy = rnorm(iT)
    vEps[30001:32000] = rnorm(2000 ,sd =seq(from = dSigma , 
                                            to = 2*dSigma , length.out = 2000)) 
    vEps[32001:34000] = rnorm(2000 ,sd =seq(from = 2*dSigma , 
                                            to = dSigma , length.out = 2000))
    vEpsy[30001:32000] = -rnorm(2000 ,mean =seq(from = 0,
                                                to = 0.3 , length.out = 2000)) 
    vEpsy[32001:34000] = -rnorm(2000 ,mean =seq(from = 0.3,
                                                to = 0 , length.out = 2000))
    
    
    vEps[60001:63000] = rnorm(3000,sd = seq(from = dSigma , 
                                            to = 2*dSigma , length.out = 3000))
    vEps[63001:66000] = rnorm(3000,  sd = seq(from = 2*dSigma , 
                                              to =  dSigma, length.out = 3000))
    
    vEpsy[60001:63000] = rnorm(3000 ,mean =seq(from = 0,
                                               to = 0.2 , length.out = 3000))
    vEpsy[63001:66000] = rnorm(3000 ,mean =seq(from = 0.2,
                                               to = 0 , length.out = 3000))
    vSim[1] = dMu + dPhi *rnorm(1 , mean = dMu , sd = dSigma /sqrt(1-dPhi^2))
    for (i in 2:iT) {
      vSim[i] = dMu + dPhi * (vSim[(i-1)] - dMu) + vEps[i]
    }
    vY = exp(vSim/2) * vEpsy
    return(vY)
  }
  #Set parameter values of the simulation
  iT = 66500; dSigma = 0.3; dPhi = 0.98; dMu = -10;
  #set seed for reproducibility
  set.seed(123)
  #Simulate the series
  vY = 500+cumsum(flashCrashSim(iT, dSigma, dPhi, dMu))
  
  #insert an outlier to illustrate robustness.
  vY[50000] = 500
  
  #Here, the observations are equidistant, but the code can handle unevenly spaced observations.
  timestamps = seq(34200 , 57600 , length.out = iT)
  testtimes = seq(34260, 57600, 60)
  logprices = log(vY)
  
  library("DriftBurstHypothesis")
  
  #calculating drift burst hypothesis
  
  DBH = driftBursts(timestamps,  logprices,
                     testtimes, preAverage = 5, ACLag = -1L,
                     meanBandwidth = 300L, varianceBandwidth = 900L,
                     parallelize = FALSE)
  
  
  #plot test statistic
  plot = plot(DBH)
  #plot both test statistic and price
  plot2 = plot(DBH, price = vY, timestamps = timestamps)
  #Plot the mu series
  plot3 = plot(DBH, which = "Mu")
  #plot the sigma series
  plot4 = plot(DBH, which = "Sigma")
  expect_equal(plot3$usr[1:2], plot4$usr[1:2])
  #plot both
  plot5 = plot(DBH, which = c("Mu", "Sigma"))
  
  #Means of the tstat, sigma, and mu series.
  expect_equal(mean(getDB(DBH)), 0.01863012)
  expect_equal(mean(getSigma(DBH)), 0.007197401)
  expect_equal(mean(getMu(DBH)), -0.000008112) 
  
  
  
  
  ################## same example with xts object:
  suppressPackageStartupMessages(library("xts"))
  #Set parameter values of the simulation
  iT = 66500; dSigma = 0.3; dPhi = 0.98; dMu = -10;
  #set seed for reproducibility
  set.seed(123)
  #Simulate the series
  vY = 500+cumsum(flashCrashSim(iT, dSigma, dPhi, dMu))
  
  #insert an outlier to illustrate robustness.
  vY[50000] = 500
  
  #Here, the observations are equidistant, but the code can handle unevenly spaced observations.
  timestamps = seq(34200 , 57600 , length.out = iT)
  StartTime = strptime("1970-01-01 00:00:00.0000", "%Y-%m-%d %H:%M:%OS", tz = "GMT")
  Tradetime = StartTime + timestamps
  testTimes = seq(34260, 57600, 60)
  
  
  price = xts(vY, Tradetime)
  
  
  DBHxts = driftBursts(timestamps = NULL,  log(price), 
                       testTimes, preAverage = 5, ACLag = -1L,
                       meanBandwidth = 300L, varianceBandwidth = 900L, 
                       parallelize = FALSE)
  plot6 = plot(DBHxts)
  expect_true(all.equal.list(plot, plot6))
  plot7 = plot(DBHxts, price = price)
  expect_true(all.equal.list(plot2, plot7))
  #check for equality
  expect_true(all.equal(as.numeric(getDB(DBH)), as.numeric(getDB(DBHxts))))
})

R/utils.R
#Utility functions for the DriftBurstHypothesis R package

### polyfit2d
fit2d3rdDegree = function(A, X, Y){
  
  X = matrix(rep(as.numeric(X),4), ncol = 4)
  Y = matrix(Y, ncol = 16, nrow = length(Y), byrow = FALSE)
  Z = matrix(1, ncol = ncol(Y), nrow = nrow(Y))
  
  for (i in 1:3) {
    Z[,1:i] = Z[,1:i] * X[,1:i]
  }
  
  for (i in 1:3) {
    
    foo = 4*i
    Z[, 1:foo] = Z[, 1:foo] * Y[, 1:foo]
    for (j in 1:3) {
      Z[, (foo+1):(foo+j)] = Z[, (foo+1):(foo+j)] * X[,1:j]
    }
  }
  p = mldivide(Z,as.numeric(A))
  return(p)
}






isDriftBurst = function(tStat, alpha){
  ##load arr:
  alpha_used = alpha
  
  nObs = length(tStat)
  rho = DBHSysData$rho
  alpha = DBHSysData$alpha
  m = DBHSysData$m
  arr = array(DBHSysData$arr, dim = c(23, 14, 6), dimnames = list('m' = m, 'rho' = rho,'alpha' = alpha))  
  
  rho_used = acf(tStat, plot = FALSE)$acf[2]
  A = matrix(NA, ncol = nrow(arr), nrow = nrow(arr))
  P = matrix(NA, ncol = dim(arr)[3], nrow = 16)
  rho = as.numeric(dimnames(arr)$rho)
  add_rho = c(0.75,0.85,0.91,0.92,0.93,0.94,0.96,0.97,0.98)
  rho_i = sort(c(rho, add_rho))
  logRho = log(1-rho)
  logRho_i = log(1-rho_i)
  logM = matrix(log(as.numeric(dimnames(arr)$m)), ncol = nrow(arr), nrow = nrow(arr), byrow = TRUE)
  logRho_mat = matrix(logRho_i, ncol = nrow(arr), nrow = nrow(arr), byrow = FALSE)
  
  for (i in 1:length(alpha)) {
    
    for (j in 1:nrow(arr)) {
      A[ ,j] = approx(logRho, arr[j, ,i], logRho_i)$y
    }
    P[,i] = fit2d3rdDegree(A, logM, logRho_mat)
  }
  
  idx = alpha %in% alpha_used
  normalizedQuantile = P[1,idx]
  for (i in 1:3) {
    normalizedQuantile = normalizedQuantile * log(nObs) + P[1+i, idx]
  }
  
  
  for (i in 1:3) {
    foo = 4 * i+1
    g = P[foo, idx]
    
    for (j in 1:3) {
      g = g * log(nObs) + P[j+ foo, idx]
    }
    normalizedQuantile = normalizedQuantile * log(1 - rho_used) + g
    
  }
  
  AM = sqrt(2 * log(nObs))
  BM = AM - 0.5 * log(pi * log(nObs)) / AM
  
  quantile = normalizedQuantile / AM + BM
  
  return(list("normalizedQuantile" = normalizedQuantile, "quantile" = quantile))
}





src/utils.cpp
#include <RcppArmadillo.h>

//This code is taken from the repository: https://github.com/coatless/r-to-armadillo which is a collection of R functions written in C++

//[[Rcpp::export]]
arma::vec cfilter(arma::vec x, arma::vec filter, int sides = 2, bool circular = false)
{
  
  int nx = x.n_elem;
  int nf = filter.n_elem;
  int nshift;
  
  if(sides == NA_INTEGER || circular == NA_LOGICAL)  Rcpp::stop("invalid input");
  
  double z, tmp;
  
  if(sides == 2){
    nshift = nf /2;
  }
  else{
    nshift = 0;
  }
  
  arma::vec out = arma::zeros<arma::vec>(nx);
  
  if(!circular) {
    for(int i = 0; i < nx; i++) {
      z = 0;
      if(i + nshift - (nf - 1) < 0 || i + nshift >= nx) {
        out(i) = NA_REAL;
        continue;
      }
      for(int j = std::max(0, nshift + i - nx); j < std::min(nf, i + nshift + 1) ; j++) {
        tmp = x(i + nshift - j);
        z += filter(j) * tmp;
      }
      out(i) = z;
    }
  } else { /* circular */
    for(int i = 0; i < nx; i++)
    {
      z = 0;
      for(int j = 0; j < nf; j++) {
        int ii = i + nshift - j;
        if(ii < 0) ii += nx;
        if(ii >= nx) ii -= nx;
        tmp = x(ii);
        z += filter(j) * tmp;
      }
      out(i) = z;
    }
  }
  return out;
}


R/RcppExports.R
# Generated by using Rcpp::compileAttributes() -> do not edit by hand
# Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

HACWeightC <- function(iLag) {
    .Call(`_DriftBurstHypothesis_HACWeightC`, iLag)
}

AsymptoticVarianceC <- function(vIn, iLag) {
    .Call(`_DriftBurstHypothesis_AsymptoticVarianceC`, vIn, iLag)
}

AutomaticLagSelectionC <- function(vX, dMu) {
    .Call(`_DriftBurstHypothesis_AutomaticLagSelectionC`, vX, dMu)
}

DriftBurstLoopC <- function(vPreAveraged, diffedlogprices, vTime, vTesttime, iMeanBandwidth, iVarBandwidth, iPreAverage, iAcLag) {
    .Call(`_DriftBurstHypothesis_DriftBurstLoopC`, vPreAveraged, diffedlogprices, vTime, vTesttime, iMeanBandwidth, iVarBandwidth, iPreAverage, iAcLag)
}

DriftBurstLoopCPAR <- function(vPreAveraged, diffedlogprices, vTime, vTesttime, iMeanBandwidth, iVarBandwidth, iPreAverage, iAcLag, iCores) {
    .Call(`_DriftBurstHypothesis_DriftBurstLoopCPAR`, vPreAveraged, diffedlogprices, vTime, vTesttime, iMeanBandwidth, iVarBandwidth, iPreAverage, iAcLag, iCores)
}

DriftBurstRealTimeC <- function(vLogPrices, diffedlogprices, vTime, vTesttime, iMeanBandwidth, iVarBandwidth, iPreAverage, iAcLag) {
    .Call(`_DriftBurstHypothesis_DriftBurstRealTimeC`, vLogPrices, diffedlogprices, vTime, vTesttime, iMeanBandwidth, iVarBandwidth, iPreAverage, iAcLag)
}

DriftBurstRealTimeCPAR <- function(vLogPrices, diffedlogprices, vTime, vTesttime, iMeanBandwidth, iVarBandwidth, iPreAverage, iAcLag, iCores) {
    .Call(`_DriftBurstHypothesis_DriftBurstRealTimeCPAR`, vLogPrices, diffedlogprices, vTime, vTesttime, iMeanBandwidth, iVarBandwidth, iPreAverage, iAcLag, iCores)
}

cfilter <- function(x, filter, sides = 2L, circular = FALSE) {
    .Call(`_DriftBurstHypothesis_cfilter`, x, filter, sides, circular)
}



R/ClassMethods.R
plot.DBH = function(x, ...){
  
  #### Get extra passed options and data
  options = list(...)
  #### List of standard options
  opt = list(which = "driftbursts", price = NULL, timestamps = NULL, startTime = 34200, endTime = 57600,
             leg.x = "topleft", leg.y = NULL,  tz = "GMT", annualize = FALSE, nDays = 252, legend.txt = "")
  #### Override standard options where user passed new options
  opt[names(options)] = options
  # for (i in 1:length(opt)) {
  #   assign(names(opt)[i], opt[[i]])
  # }

  #### Extract options (better way to do it?)  #I will test above method when I get time.
  which      = tolower(opt$which)
  startTime  = opt$startTime
  endTime    = opt$endTime
  main       = opt$main
  tz         = opt$tz
  leg.x      = opt$leg.x
  leg.y      = opt$leg.y
  annualize  = opt$annualize
  nDays      = opt$nDays
  price      = opt$price
  timestamps = opt$timestamps
  tstat      = getDB(x)
  sigma      = getSigma(x, annualize, nDays) 
  mu         = getMu(x, annualize, nDays)
  startpar   = par(no.readonly = TRUE)
  testTimes  = x$info$testTimes
  horizLines = seq(round(min(tstat)), round(max(tstat)), 1)
  ###Setup done
  if(!all(which %in% c("driftbursts", "mu", "sigma", "db"))){
    stop("The which argument must be a character vector containing either:\n
         Sigma, Mu, both of these or DriftBursts. 
         CasE doesn't matter.")
  }
  if(inherits(tstat, "xts")){
    testTimes = index(tstat)
    testTimes = as.numeric(testTimes) - (.indexDate(tstat)[1] * 86400)
    tstat     = as.numeric(tstat)
    sigma     = as.numeric(sigma)
    mu        = as.numeric(mu)
    if(!is.null(price)){
      tz          = tzone(price)
      timestamps  = index(price, tz = tz)
      timestamps  = as.numeric(timestamps) - (.indexDate(price)[1] * 86400)
      price       = as.numeric(price)
    }
  }
  if(testTimes[1] == startTime){
    testTimes = testTimes[-1]
    sigma = sigma[-1]
    mu = mu[-1]
    tstat=tstat[-1]
  }
  if(min(testTimes) < startTime | max(testTimes) > endTime){
    cat('\nTesting was tried before startTime or after endTime, thus some of the tests may be cut off from the plot.
        \nIf the plot looks weird, consider changing startTime and endTime. 
        \nThese should reflect the start of trading and the end of trading respectively')
  }
  xtext = as.POSIXct(testTimes, origin = "1970-01-01", tz = tz)
  xlim = c(startTime, endTime)
  xlab = "Time"
  
  if(all(which %in% c("driftbursts", "db"))){ #use all() because this function should accept which arguments with length longer than 1
    par(mar = c(4,3.5,2,1.25), mgp = c(2,1,0))
    if(!is.null(price)) par(mar = c(4,3.5,4,4), mgp = c(2,1,0)) #makes room for values on the right y-axis
    main = "Drift Bursts test statistic"
    ylab = "test-statistic"
    plot(tstat, x = xtext, type = "l", xaxt = 'n', ylab = ylab, main = main, xlab = xlab, xlim = xlim)
    axis.POSIXct(side  = 1, at = seq.POSIXt(xtext[1], xtext[length(xtext)], length.out = 7))
    abline(h = horizLines, col = "grey" , lty = 3, cex = 0.1)
    legend.txt = "t-stat"
    if(!is.null(price)){
      if(is.null(timestamps)){
        stop("The timestamps of the price must be passed in the timestamps argument")
      }
      par(new = TRUE)
      plot(price, x = timestamps , type = "l", axes = FALSE, col = "red", xlab = "", ylab = "", lty = 2, xlim = xlim)  
      axis(4)
      mtext(side = 4, text = "price", line = 2.5)
      legend.txt = c(legend.txt, "price")
      legend(x = leg.x, leg.y, legend = legend.txt, lty = c(1,2), col = c(1,2), bg = rgb(0,0,0,0), box.lwd = 0, 
             box.col = rgb(0,0,0,0))
    }
  }
  if(all(which == "sigma")){ #use all() because this function should accept which arguments with length longer than 1
    main = "volatility"
    ylab = "local volatility"
    par(mar = c(4,3.5,2,1.25), mgp = c(2,1,0))
    plot(sigma, x = xtext, type = "l",  xaxt = 'n', ylab = ylab, main = main, xlab = xlab)  
    axis.POSIXct(side  = 1, at = seq.POSIXt(xtext[1], xtext[length(xtext)], length.out = 7))
  }
  if(all(which == "mu")){ #use all() because this function should accept which arguments with length longer than 1
    main = "drift"
    ylab = "drift"
    if(annualize){ 
      ylab = paste("annualized", ylab)
    }
    par(mar = c(4,3.5,2,1.25), mgp = c(2,1,0))
    plot(mu, x = xtext, type = "l",  xaxt = 'n', ylab = ylab, main = main, xlab = xlab)
    axis.POSIXct(side  = 1, at = seq.POSIXt(xtext[1], xtext[length(xtext)], length.out = 7))
    abline(h = 0, col = "grey" , lty = 3)
  }
  if("mu" %in% which & "sigma" %in% which){
    par(mfrow = c(2,1), omi = c(0,0,0,0), mgp = c(2,1,0), mai = c(0.75,0.75,0.3,0.25))
    main = "drift"
    ylab = "drift"
    if(annualize){ 
      ylab = paste("annualized", ylab)
    }
    plot(mu, x = xtext, type = "l", xlab = "",  xaxt = 'n', ylab = ylab, main = main)
    axis.POSIXct(side  = 1, at = seq.POSIXt(xtext[1], xtext[length(xtext)], length.out = 7))
    abline(h = 0, col = "grey" , lty = 3)
    main = "volatility"
    ylab = "volatility"
    if(annualize){ 
      ylab = paste("annualized", ylab)
    }
    plot(sigma, x = xtext, type = "l", xlab = "", xaxt = 'n', ylab = ylab, main = main) 
    axis.POSIXct(side  = 1, at = seq.POSIXt(xtext[1], xtext[length(xtext)], length.out = 7))
  }
  par(startpar)
}

print.DBH = function(x, ...){
  usePolynomialInterpolation = TRUE
  options = list(...)
  if('criticalValue' %in% names(options)){
    usePolynomialInterpolation = FALSE
  }
  #### List of standard options
  opt = list(alpha = 0.95)
  #### Override standard options where user passed new options
  opt[names(options)] = options
  if(usePolynomialInterpolation){
    alpha = opt$alpha
    criticalValue = isDriftBurst(getDB(x), alpha)$quantile  
  }else{
    criticalValue = opt$criticalValue
  }
  
  varDB = getVar(x, which = 'db')
  padding = x$info$padding
  #We always remove the first entry, as this is used to denote the start of trading.
  whichToInclude = seq(padding[1] + 1, length(x$info$testTimes)- padding[2]) 
  allMeans = getMean(x, which = 'all')
  cat("\n-------------Drift Burst Hypothesis------------\n")
  cat("Tests performed:                     ", length(whichToInclude))
  #browser()
  if(usePolynomialInterpolation){
    cat("\nAny drift bursts (|T| > ",paste0(round(criticalValue,3)),"):    ", ifelse(any(abs(getDB(x))>criticalValue) , 'yes', 'no'))
  }else{
    cat("\nAny drift bursts (|T| > ",paste0(criticalValue[1]),"):        ", ifelse(any(abs(getDB(x))>criticalValue) , 'yes', 'no'))
  }
  cat("\nMax absolute value of test statistic:", round(max(abs(getDB(x))), digits=5))
  cat("\nMean test statistic:                 ", round(allMeans$meanDB, digits = 5))
  cat("\nVariance of test statistic:          ", round(varDB, digits = 5))
  cat("\nMean mu:                             ", round(allMeans$meanMu, digits = 5))
  cat("\nMean sigma:                          ", round(allMeans$meanSigma, digits = 5))
  cat("\n-----------------------------------------------\n")
}

getDB = function(x){
  UseMethod("getDB", x)
}

getDB.DBH = function(x){
  DB = x$driftBursts
  return(DB)
}


getSigma = function(x, annualize = FALSE, nDays = 252){
  UseMethod("getSigma", x)
}

getSigma.DBH = function(x, annualize = FALSE, nDays = 252){
  sigma = sqrt((x$sigma * 2 * x$info$nObs)  / (x$info$nObs / 23400))/(x$info$preAverage^2)
  if(annualize){sigma = sigma * sqrt(nDays)}
  return(sigma)
}


getMu = function(x, annualize = FALSE, nDays = 252){
  UseMethod("getMu", x)
}

getMu.DBH = function(x, annualize = FALSE, nDays = 252){
  mu = (x$mu * x$info$meanBandwidth / (x$info$nObs / 23400)) / (x$info$preAverage^2 * 2)
  if(annualize){mu =  mu * nDays}
  return(mu)
}

getMean = function(x, which = 'all'){
  UseMethod('getMean', x)
}

getMean.DBH = function(x, which = 'all'){
  which     = tolower(which)
  padding = x$info$padding
  if(!(which %in% c('all', 'db', 'driftbursts', 'mu', 'sigma'))){
    stop("The which argument must be a character vector containing either:\n
         Sigma, Mu, both of these or DriftBursts. 
         CasE doesn't matter.")
  }
  #We always remove the first entry, as this is used to denote the start of trading.
  whichToInclude = seq(padding[1] + 1, length(x$info$testTimes)- padding[2]) 
  
  if(which == 'all'){
    meanDB = mean(getDB(x)[whichToInclude])
    meanMu = mean(getMu(x)[whichToInclude])
    meanSigma = mean(getSigma(x)[whichToInclude])  
    out = list('meanDB' = meanDB, 'meanMu' = meanMu, 'meanSigma' = meanSigma)
  }
  if(which %in% c('driftbursts', 'db')){
    out = mean(getDB(x)[whichToInclude])
  }
  if(which == 'mu'){
    out = mean(getMu(x)[whichToInclude])
  }
  if(which == 'sigma'){
    out = mean(getSigma(x)[whichToInclude])
  }
  
  return(out)
  
}


getVar = function(x, which = 'all', annualize = FALSE, nDays = 252){
  UseMethod('getVar', x)
}


getVar.DBH = function(x, which = 'all', annualize = FALSE, nDays = 252){
  which     = tolower(which)
  padding = x$info$padding
  if(!(which %in% c('all', 'db', 'driftbursts', 'mu', 'sigma'))){
    stop("The which argument must be a character vector containing either:\n
         Sigma, Mu, both of these or DriftBursts. 
         CasE doesn't matter.")
  }
  #We always remove the first entry, as this is used to denote the start of trading.
  whichToInclude = seq(padding[1] + 1, length(x$info$testTimes)- padding[2]) 
  if(which == 'all'){
    varDB = var(getDB(x)[whichToInclude])
    varMu = var(getMu(x, annualize, nDays)[whichToInclude])
    varSigma = var(getSigma(x, annualize, nDays)[whichToInclude])  
    out = list('varDB' = varDB, 'varMu' = varMu, 'varSigma' = varSigma)
  }
  if(which %in% c('driftbursts', 'db')){
    out = var(getDB(x)[whichToInclude])
  }
  if(which == 'mu'){
    out = var(getMu(x, annualize, nDays)[whichToInclude])
  }
  if(which == 'sigma'){
    out = var(getSigma(x, annualize, nDays)[whichToInclude])
  }
  
  return(out)

}